src/storage/storageController.ts

```diff
@@ -24,52 +26,50 @@ export enum DefaultSavingLocations {
   'Session Only' = 'Session Only'
 }
 
-// A saved connection does not contain the connection information,
-// just metadata about the connection and an id for referencing secure storage.
-export type SavedConnection = {
-  id: string; // uuidv4
-  name: string; // Possibly user given name, not unique.
-  storageLocation: StorageScope;
+export type ConnectionsFromStorage = {
+  [key: string]: StoreConnectionInfo
 };
 
-type StoredConnectionsType = { [key: string]: SavedConnection } | undefined;
+type StoredVariableName = StorageVariables.GLOBAL_USER_ID |
+  StorageVariables.GLOBAL_SAVED_CONNECTIONS |
+  StorageVariables.WORKSPACE_SAVED_CONNECTIONS |
+  StorageVariables.GLOBAL_HAS_BEEN_SHOWN_INITIAL_VIEW;
+
+type StoredItem<T> =
+    T extends StorageVariables.GLOBAL_USER_ID ? string :
+    T extends StorageVariables.GLOBAL_HAS_BEEN_SHOWN_INITIAL_VIEW ? boolean :
+    T extends StorageVariables.GLOBAL_SAVED_CONNECTIONS ? ConnectionsFromStorage :
+    T extends StorageVariables.WORKSPACE_SAVED_CONNECTIONS ? ConnectionsFromStorage :
+    never;
 
 export default class StorageController {
-  _globalState: vscode.Memento;
-  _workspaceState: vscode.Memento;
+  _storage: { [StorageLocation.GLOBAL]: vscode.Memento, [StorageLocation.WORKSPACE]: vscode.Memento };
 
   constructor(context: vscode.ExtensionContext) {
-    this._globalState = context.globalState;
-    this._workspaceState = context.workspaceState;
+    this._storage = {
+      [StorageLocation.GLOBAL]: context.globalState,
+      [StorageLocation.WORKSPACE]: context.workspaceState
+    };
   }
 
-  get(variableName: StorageVariables, storageScope?: StorageScope): any {
-    if (storageScope === StorageScope.WORKSPACE) {
-      return this._workspaceState.get(variableName);
-    }
-
-    return this._globalState.get(variableName);
+  get<T extends StoredVariableName>(variableName: T, storageLocation: StorageLocation = StorageLocation.GLOBAL): StoredItem<T> {
+    return this._storage[storageLocation].get(variableName);
   }
 
   // Update something in the storage. Defaults to global storage (not workspace).
   update(
     variableName: StorageVariables,
-    value: any,
-    storageScope?: StorageScope
+    value: boolean | string | ConnectionsFromStorage,
+    storageLocation: StorageLocation = StorageLocation.GLOBAL
   ): Thenable<void> {
```

Comment:
(nit:) This could also be

```js

  update<T extends StoredVariableName>(
    variableName: T,
    value: StoredItem<T>,
    storageLocation: StorageLocation = StorageLocation.GLOBAL
  ): Thenable<void>
```

to make sure that `variableName` and `value` match